### ISSUE TITLE
model: support bounded recurrent-state rollback for Kimi-K3

### ISSUE THREAD

**Issue by soulmachine** [ggml-org/llama.cpp#28461]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

## Requested behavior

Support bounded recurrent-state rollback for Kimi-K3 so speculative decoding can restore recurrent state at the accepted token boundary. Host checkpoints already provide a fallback; bounded snapshots could reduce that overhead.

At commit `74a7c897f049c17e7080423aa2111776eff6ebbf`, Kimi-K3 is excluded from `llm_arch_supports_rs_rollback`, and a requested nonzero `n_rs_seq` is clamped to zero. Its KDA layer stores only the final recurrent state and Q/K/V convolution windows.

Simply adding Kimi-K3 to the allowlist leaves historical snapshot groups unwritten. The diagnostic below exposes that gap. Stock Kimi-K3 has rollback disabled; the failing diagnostic requires both force-enabling rollback and initializing the recurrent cache with a nonzero byte pattern.

## Reproducer

Environment used for the recorded results: macOS, Apple clang, CPU-only Release build, and the generated dummy `kimi-k3-moe.gguf`. No real model weights are required.

Start from the pinned revision and apply two diagnostic changes: enable Kimi-K3 in the rollback allowlist, then add an environment guard for nonzero cache initialization. The second replacement changes only the constructor's buffer clear; the later clear path stays unchanged.

```sh
git clone https://github.com/ggml-org/llama.cpp.git llama-k3-rollback-repro
cd llama-k3-rollback-repro
git checkout 74a7c897f049c17e7080423aa2111776eff6ebbf

python3 - <<'PY'
from pathlib import Path

path = Path('src/llama-arch.cpp')
source = path.read_text()
old = 'bool llm_arch_supports_rs_rollback(const llm_arch & arch) {\n    switch (arch) {\n'
assert source.count(old) == 1
path.write_text(source.replace(old, old + '        case LLM_ARCH_KIMI_K3:\n', 1))

path = Path('src/llama-memory-recurrent.cpp')
source = path.read_text()
if '#include <cstdlib>' not in source:
    assert '#include <cstring>' in source
    source = source.replace('#include <cstring>', '#include <cstdlib>\n#include <cstring>', 1)
old = 'ggml_backend_buffer_clear(buf, 0);'
new = 'ggml_backend_buffer_clear(buf, std::getenv("RS_POISON") ? 0x3e : 0);'
assert old in source
path.write_text(source.replace(old, new, 1))
PY

cmake -B build-cpu -DGGML_METAL=OFF -DGGML_BLAS=OFF -DLLAMA_CURL=OFF \
      -DLLAMA_BUILD_TESTS=ON -DCMAKE_BUILD_TYPE=Release
cmake --build build-cpu -j6 --target test-llama-archs test-recurrent-state-rollback
mkdir -p models-generated
./build-cpu/bin/test-llama-archs -o models-generated/

env -u RS_POISON ./build-cpu/bin/test-recurrent-state-rollback \
    -m models-generated/kimi-k3-moe.gguf
RS_POISON=1 ./build-cpu/bin/test-recurrent-state-rollback \
    -m models-generated/kimi-k3-moe.gguf
```

`RS_POISON` is a temporary diagnostic guard added by the snippet, not an upstream option. Its presence fills the constructor's buffer with byte `0x3e`, producing the finite float32 pattern `0x3e3e3e3e` (about `0.18578431`). With the variable unset, initialization remains zero-filled.

## Recorded results

The existing `test_multi_seq_split_replay` uses two sequences, a 16-token microbatch, a 19-token prompt, rollback by three tokens, and a 40-token replay. Its comparison tolerance is `1e-7`.

| Kimi-K3 implementation | Zero-filled cache | Nonzero diagnostic cache |
| --- | --- | --- |
| Pinned revision plus allowlist entry only | Pass, max difference `3.66708e-09` | Fail, max difference `7.67533e-05`; first mismatch at sequence 0, position 16 |
| Same revision plus KDA and convolution snapshot stores | Pass, max difference `4.65661e-10` | Pass, max difference `4.65661e-10` |

Expected behavior is matching replay logits after restoring the accepted prefix. The failure appears when rollback reads a historical group that the original Kimi-K3 stores never populated. On the tiny dummy model, zero initialization hides the missing writes within the test tolerance.

The same nonzero-cache diagnostic passed for the already-supported generated Qwen3.5, Nemotron-H, and DeepSeek-V4 models, with multi-sequence replay differences of zero. These are recorded results from the earlier diagnostic; the temporary cache-initialization edit was subsequently reverted.

### Motivation

Support bounded recurrent-state rollback for Kimi-K3 so speculative decoding can restore recurrent state at the accepted token boundary. Host checkpoints already provide a fallback; bounded snapshots could reduce that overhead.

### Possible Implementation

Use the existing `build_recurrent_attn` helper for KDA state snapshots and store each Q/K/V convolution window at the corresponding rollback position. Enable the architecture in the rollback allowlist once both stores populate the snapshot groups. A local two-file patch implements this and produced the passing results above.

The stock dummy-model test alone does not detect the missing writes; improving that coverage remains a separate consideration. Full CI and other backends were not covered by this CPU diagnostic.

Related: [#28019](https://github.com/ggml-org/llama.cpp/issues/28019) reports rollback corruption for qwen4exp. Similar missing snapshot coverage is a possible connection, inferred from the two cases; this report concerns Kimi-K3.

AI usage disclosure: YES - AI assisted with implementation, diagnosis, and validation. Codex prepared this report from the source and recorded diagnostic results.
